core/internal/api/credentials.go

```diff
@@ -55,3 +65,237 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
 	)
 	return nil
 }
+
+var _ CredentialProvider = &oauth2CredentialProvider{}
+
+// OAuth2CredentialProvider creates a credentials provider that exchanges a JWT
+// for an access token via an authorization server. The access token is then used
+// to authenticate API requests.
+//
+// The JWT is supplied via a file path that is passed in as an environment
+// variable. When the OAuth2CredentialProvider is applied, it exchanges the JWT
+// for an access token. Once the token is received, it is saved to the
+// credentials file along with its expiration. The expiration is checked each
+// time the access token is used, and refreshed if it is at or near expiration.
+func NewOAuth2CredentialProvider(
+	settings *settings.Settings,
+) (CredentialProvider, error) {
+	return &oauth2CredentialProvider{
+		baseURL:               settings.GetBaseURL(),
+		identityTokenFilePath: settings.GetIdentityTokenFile(),
+		credentialsFilePath:   settings.GetCredentialsFile(),
+		mu:                    &sync.Mutex{},
+	}, nil
+}
+
+type oauth2CredentialProvider struct {
+	// The URL of the W&B API
+	baseURL string
+
+	// The file path to the JWT
+	identityTokenFilePath string
+
+	// The file path to the access token and its metadata
+	credentialsFilePath string
+
+	// The access token and its metadata
+	token tokenInfo
+
+	mu *sync.Mutex
+}
+
+// ExpiresAt is a custom type representing a time.Time value. It is used to handle
+// expiration times in a specific string format when serializing/deserializing JSON data.
+type ExpiresAt time.Time
+
+const expiresAtLayout = "2006-01-02 15:04:05"
+
+func (e *ExpiresAt) UnmarshalJSON(data []byte) error {
+	var timeString string
+	if err := json.Unmarshal(data, &timeString); err != nil {
+		return err
+	}
+
+	parsedTime, err := time.Parse(expiresAtLayout, timeString)
+	if err != nil {
+		return err
+	}
+
+	*e = ExpiresAt(parsedTime)
+	return nil
+}
+
+func (e ExpiresAt) MarshalJSON() ([]byte, error) {
+	formattedTime := time.Time(e).Format(expiresAtLayout)
+	return json.Marshal(formattedTime)
+}
+
+type tokenInfo struct {
+	// The time at which the token will expire
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *tokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]tokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.token.IsTokenExpiring() {
+		if err := c.loadCredentials(); err != nil {
+			return err
+		}
+	}
```

Comment:
Since Apply can get called by multiple goroutines concurrently, this is a data race: `c.loadCredentials` can modify `c.token` while another `Apply` reads it.

Another issue is that multiple `Apply` simultaneous calls that all see an expired token will *all* go through the `loadCredentials` path one after another, even though `loadCredentials` should only happen once.

You can use a read-write lock to accomplish this without a lot of overhead. The happy path uses the very fast `RLock()`; the expired-token path uses a slow write `Lock()` but skips doing work if the token is already fresh. Here's a sketch:

```go
type oauth2CredentialProvider struct {
	token   tokenInfo
	tokenMu sync.RWMutex
}

func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
	token, err := c.freshToken()
	if err != nil {
		return err
	}

	req.Header.Set(
		"Authorization",
		"Bearer "+token.AccessToken,
	)

	return nil
}

func (c *oauth2CredentialProvider) freshToken() (tokenInfo, error) {
	if token, ok := c.isTokenFresh(); ok {
		return token, nil
	}

	return c.refreshToken()
}

func (c *oauth2CredentialProvider) isTokenFresh() (tokenInfo, bool) {
	c.tokenMu.RLock()
	defer c.tokenMu.RUnlock()

	if c.token.IsTokenExpiring() {
		return tokenInfo{}, false
	}

	return c.token, true
}

func (c *oauth2CredentialProvider) refreshToken() (tokenInfo, error) {
	c.tokenMu.Lock()
	defer c.tokenMu.Unlock()

	if !c.token.IsTokenExpiring() {
		return c.token, nil
	}

	... // the regular logic
}
```

Reply:
I had a lock on the loadCredentials function, so we would not re-request the token. In any case, I've updated to this approach.